### PR TITLE
fix(email-ingestion): classify forwarded, relayed and self-issued mail

### PR DESCRIPTION
**Pull request by gilgardosh**
## Problem

Issuer recognition held for mail sent straight to the tenant, but collapsed on the two shapes that dominate real inboxes.

**Manual forwards were misfiled as self-issued.** A hand forward rewrites `From` to the forwarder, drops the original `Reply-To` and sets no `X-Original-*`. When the quoted headers are themselves a mailing-list rewrite, *every* recoverable address belongs to the tenant — so recognition matched the tenant's own business and ingest withheld the document as self-issued. A forwarded Mailchimp invoice was indistinguishable from a genuine self-issued one.

**Self-issued detection worked by accident.** It fired only because `INVOICE_ISSUING_PROVIDER_EMAILS` hard-coded `ap@the-guild.dev`, one specific tenant's forwarding group. No other tenant had any detection at all.

## Approach

`selectIssuerEmail` / `selectIssuerCandidates` / `isSelfIssuedSenderEvidence` are replaced by a single `classifyEmail(evidence, tenantContext)` returning `{ kind, issuerCandidates, forwarder, issuerNameHint }`. Kind is first-match-wins:

| # | condition | kind |
| --- | --- | --- |
| 1 | a quoted forwarded block exists, or `From` is a person at the tenant | `FORWARDED` |
| 2 | the sender's display name is one of the tenant's own business names | `SELF_ISSUED` |
| 3 | came through an invoice platform **and** an external address survives | `RELAYED` |
| 4 | came through an invoice platform and none does | `SELF_ISSUED` |
| 5 | a mailing-list marker or a `'X' via Y` display name is present | `RELAYED` |
| 6 | otherwise | `DIRECT` |

Rule 1 outranking rule 4 is the crux: a person deliberately forwarding into the ingest alias signals intent to ingest, whereas self-issued confirmations always arrive by automatic relay, never by hand.

Candidates are tiered (innermost quoted `From` → `Reply-To` → `X-Original-From` → `From` → body `mailto:`, invoice-platform addresses last), and the tenant's own addresses, its mailing-list addresses and the forwarder are excluded at every tier. `issuerNameHint` carries the sender's display name for the case where a list rewrite leaves no usable address at all — that is what the name-based matcher and, ultimately, OCR work from.

**Tenant context** comes from the tenant's ingest aliases plus its *own* business row — deliberately `b.id = tenant`, not `b.owner_id = tenant`, since the latter is every counterparty in the workspace — plus explicit `emailIngestion.ownDomains` / `extraPlatformSenders` config for colleagues registered nowhere.

**Outcome:** self-issued mail is now `IngestOutcome.IGNORED` rather than `QUARANTINED` — still recorded and reprocessable, but no longer filling the operator triage queue. The signal is a new `email_ingestion_grants.classification` column; grants issued before it fall back to the old `business_id = owner_id` comparison.

**Gateway:** new `forwarded.ts` parses nested quoted forward blocks across both the text and HTML parts, and ships `fromDisplayName` / `originalSender` / `listId` / `listAddresses` / `forwardedBlocks` as additive sender evidence. It holds no tenant knowledge.

## Two defects found while building

- `addressParser` strips `(via Paddle.com)` as an RFC 5322 comment, so quoted display names are now sliced rather than parsed.
- `headerValue` returned the first match, letting `X-Original-From` shadow `X-Original-Sender` — the single most reliable "came through an invoice platform" signal. They are separate fields now.

## Verification

Checked against eight real captured messages (kept local, never committed):

| sample | kind | issuer candidates / hint |
| --- | --- | --- |
| 2 × platform-relayed supplier invoice | `RELAYED` | the real vendor address |
| 3 × self-issued copy | `SELF_ISSUED` | — (skipped, not quarantined) |
| nested double forward | `FORWARDED` | Stripe + `support@transistor.fm`, hint `Transistor.fm` |
| forwarded Mailchimp invoice | `FORWARDED` | none, hint `No Reply - Mailchimp` |
| forwarded Paddle receipt | `FORWARDED` | `help@paddle.com`, hint `screeenly (via Paddle.com)` |

The nested forward now recovers `Transistor.fm` as the issuer name, which the old code lost entirely.

- 460 tests pass across the touched packages; `yarn lint`, `prettier:check` and both typechecks are clean.
- Full suite and integration baselines are unchanged: 18 integration suites and `rls-all-tables.test.ts` fail identically on a clean tree (local DB permissions), with 17 more tests passing after this change and zero new failures.
- Committed fixtures are hand-authored with fictional data, one per real-world header shape. `packages/email-ingestion-gateway/example-docs/` was **not** git-ignored — real vendor addresses, DKIM signatures and real invoice PDFs, one `git add -A` from being committed. It is ignored now.

## Rollout — order matters

GraphQL rejects unknown input fields at validation, so **the server must be live before the gateway ships**; otherwise `server-client` maps the error to `TRANSIENT_UPSTREAM` and retries, stalling all inbound mail. The reverse is safe: an upgraded server receiving old evidence degrades to header-only classification.

`ap@the-guild.dev` stays in `GLOBAL_INVOICE_PLATFORM_SENDERS` until that tenant's `emailIngestion.ownDomains` is populated and verified in production — then remove it (tracked in the plan doc).

## Out of scope

Three real samples still yield zero documents: the invoice sits behind a body link, and link-fetch config is only returned once a business is recognized. Recorded as a follow-up in `docs/multi-tenant-gmail-listener/business-recognition-plan.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/server/src/modules/email-ingestion/providers/email-ingestion-control.provider.ts [hunk 1]
```diff
@@ -265,28 +309,80 @@ export class EmailIngestionControlProvider {
...
+        invoicePlatformSenders: new Set(
+          [...GLOBAL_INVOICE_PLATFORM_SENDERS, ...(config.extraPlatformSenders ?? [])]
+            .map(sender => normalizeEmail(sender))
+            .filter((sender): sender is string => sender !== undefined),
+        ),
+      } satisfies TenantMailContext;
+    });
+
+    this.mailContextCache.set(tenantId, { context, expiresAt: Date.now() + MAIL_CONTEXT_TTL_MS });
+    return context;
```
[the remark this answers is not in the document]
Comment (gilgardosh) [reply packages/server/src/modules/email-ingestion/providers/email-ingestion-control.provider.ts:396]: Fixed in a9989e6a4. Expired entries are now swept before each insert, so the map is bounded by the tenants active within one TTL window rather than every tenant that ever received mail. The sweep only runs on a cache miss (at most once per tenant per TTL), so it stays off the hot path. Covered by a test that loads three tenants, advances past the TTL, then loads a fourth and asserts only the fourth remains.

File: packages/email-ingestion-gateway/src/forwarded.ts [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+function decodeEntities(value: string): string {
+  return value.replace(/&(#x?[0-9a-f]+|[a-z]+);/gi, (match, entity: string) => {
+    if (entity.startsWith('#')) {
+      const codePoint =
+        entity[1]?.toLowerCase() === 'x'
+          ? Number.parseInt(entity.slice(2), 16)
+          : Number.parseInt(entity.slice(1), 10);
+      // Zero-width and narrow no-break spaces are common in Gmail's quoted blocks
+      // (e.g. `&#8239;` before AM/PM) and must not become literal odd characters.
+      return Number.isFinite(codePoint) ? String.fromCodePoint(codePoint) : match;
```
[the remark this answers is not in the document]
Comment (gilgardosh) [reply packages/email-ingestion-gateway/src/forwarded.ts]: Good catch — the comment described a normalization the code did not perform. Fixed in a9989e6a4: space-like code points (U+00A0, U+202F, U+205F, U+2007, U+3000) fold to a regular space and zero-width ones (U+200B–U+200D, U+2060, U+FEFF) are dropped.

The concrete failure mode is narrower but real: an invisible character inside an address makes it stop matching `EMAIL_RE` in the server classifier, and inside a display name it stops comparing equal to the tenant's own business name — which is one of the self-issued signals.

While in there I also found `String.fromCodePoint` throws `RangeError` above U+10FFFF, which the `Number.isFinite` guard did not catch, so a malformed entity like `&#99999999;` would have thrown out of the whole parse. Out-of-range values now fall through to the literal text. Both are covered by new tests in `forwarded.test.ts`.
